Update guzzle (7.6.1 => 7.9.2)

eileenmcnaughton opened the pull request:
Overview
----------------------------------------
Update guzzle (7.6.1 => 7.9.2)

This will be forced on us soon by php 8.4 & our CMS updating to support it 

Note we have been happily running Guzzle 7.8.x without clashes with the core 7.6.1 - although this goes slightly higher than that

Before
----------------------------------------
7.6.1

After
----------------------------------------
7.92

Technical Details
----------------------------------------
docker@civicrm:/srv/civi-sites/dmaster/web/sites/all/modules/civicrm$ composer update -w guzzlehttp/guzzle
Gathering patches for root package.
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 7 updates, 0 removals
  - Upgrading guzzlehttp/guzzle (7.6.1 => 7.9.2)
  - Upgrading guzzlehttp/promises (1.5.1 => 2.0.3)
  - Upgrading guzzlehttp/psr7 (1.9.1 => 2.7.0)
  - Upgrading psr/http-client (1.0.1 => 1.0.3)
  - Upgrading symfony/deprecation-contracts (v2.5.2 => v2.5.3)
  - Upgrading symfony/polyfill-mbstring (v1.27.0 => v1.30.0)
  - Upgrading symfony/polyfill-php72 (v1.27.0 => v1.30.0)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 7 updates, 0 removals
  - Downloading symfony/polyfill-php72 (v1.30.0)
Gathering patches for root package.
Gathering patches for dependencies. This might take a minute.
  - Upgrading symfony/polyfill-mbstring (v1.27.0 => v1.30.0): Extracting archive
  - Upgrading symfony/deprecation-contracts (v2.5.2 => v2.5.3): Extracting archive
  - Upgrading psr/http-client (1.0.1 => 1.0.3): Extracting archive
  - Upgrading guzzlehttp/psr7 (1.9.1 => 2.7.0): Extracting archive
  - Upgrading guzzlehttp/promises (1.5.1 => 2.0.3): Extracting archive
  - Upgrading guzzlehttp/guzzle (7.6.1 => 7.9.2): Extracting archive
  - Upgrading symfony/polyfill-php72 (v1.27.0 => v1.30.0): Extracting archive

Comments
----------------------------------------

seamuslee001 commented:
This only affects what is in the tarball for Joomla, WordPress, Drupal 7, Backdrop

seamuslee001 commented:
but also tagging @artfulrobot @mattwire in case they need to update their replace statements in their compoer.jsons as needed

eileenmcnaughton commented:
Hmm it fails on a related test

![image](https://github.com/user-attachments/assets/5b2378d7-e1a1-4b05-b8c5-90bb7ec147cd)

and on a line that I don't think I would ever even start to try to debug 

![image](https://github.com/user-attachments/assets/b4bfcfd1-548e-4349-ac7b-a2c86875113d)

(annoyingly it is a fairly recent test - I wish we had updated guzzle BEFORE it was written)

eileenmcnaughton committed:
Update guzzle to 7.8.2

eileenmcnaughton commented:
I'm trying with 7.8.2 - which is what we run on WMF. 

Since guzzle upgraded to support 8.4 5 months ago 7.8.2 already has php 8.4 support so it will get us through & I'm not game to try to look at that test

docker@civicrm:/srv/civi-sites/dmaster/web/sites/all/modules/civicrm$ composer update -w guzzlehttp/guzzle:7.8.2
Gathering patches for root package.
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 5 updates, 0 removals
  - Upgrading guzzlehttp/guzzle (7.6.1 => 7.8.2)
  - Upgrading guzzlehttp/promises (1.5.1 => 2.0.3)
  - Upgrading guzzlehttp/psr7 (1.9.1 => 2.7.0)
  - Upgrading psr/http-client (1.0.1 => 1.0.3)
  - Upgrading symfony/deprecation-contracts (v2.5.2 => v2.5.3)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 3 updates, 0 removals
  - Downloading guzzlehttp/guzzle (7.8.2)
Gathering patches for root package.
Gathering patches for dependencies. This might take a minute.
  - Downgrading symfony/polyfill-mbstring (v1.30.0 => v1.27.0): Extracting archive
  - Downgrading guzzlehttp/guzzle (7.9.2 => 7.8.2): Extracting archive
  - Downgrading symfony/polyfill-php72 (v1.30.0 => v1.27.0): Extracting archive
Package laminas/laminas-zendframework-bridge is abandoned, you should avoid using it. No replacement was suggested.
Generating autoload files
30 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
> bash tools/scripts/composer/dompdf-cleanup.sh
> bash tools/scripts/composer/tcpdf-cleanup.sh
> bash tools/scripts/composer/pear-exception-fix.sh
> bash tools/scripts/composer/net-smtp-fix.sh
> bash tools/scripts/composer/pear-mail-fix.sh
> bash tools/scripts/composer/phpword-jquery.sh
> bash tools/scripts/composer/guzzle-mockhandler-fix.sh
Compiling additional files (For full details, use verbose "-v" mode.)
Compile: Generate CCL wrapper functions
Compile: Greenwich CSS (dist/bootstrap3.css)

eileenmcnaughton commented:
nope - still the same - ug

eileenmcnaughton commented:
I put up a PR to just do the psr bundle - I'm gonna try to focus on the other packages first - we know that guzzle CAN DO php 8.4 & has been able to for 5 months so perhaps the more important thing is to find which packages can't yet & push on that - e.g zeta mail I created a PR yesterday & the maintainer improved it & that package should be compliant now

totten committed:
guzzlehttp/psr7 - Switch to 1.9+ or 2.0+

* Re: guzzlehttp/psr7 symbols
    * In v1.0 - v1.8, the helper is only named `stream_for()`
    * In v1.9, the helper has two names: `stream_for()` and `Utils::streamFor()`.
    * In v2.0, the helper is only named `Utils::streamFor()`.
* This patch switches to the newer `Utils::streamFor()`. This means:
    * v1.9 and v2.0 will work.
    * v1.0 - v1.8 will *no longer* work. (This is why we need a constraint.)
* The new constraint 'guzzlehttp/psr7: ^1.9 || ^2.0' should be amenable with
  any version of Guzzle 6.x or 7.x.
    * Guzzle 6.x *requires* guzzlehttp/psr7@^1.0.0. (No conflict with ^1.9.)
    * Guzzle 7.x *varies*.
        * 7.0 requires guzzlehttp/psr7@^1.6.1. (No conflict with ^1.9.)
        * 7.1-7.8 accept a mix of in-between constraints.
        * 7.9 requires guzzlehttp/psr7@^2.7.0. (No conflict with ^2.0.)

totten commented:
The underlying failure is an internal HTTP request:

```  
      <div id="center"><div id="squeeze"><div class="right-corner"><div class="left-corner">
          <h2>Error</h2>          <div class="messages error">
<h2 class="element-invisible">Error message</h2>
<em class="placeholder">Error</em>: Call to undefined function GuzzleHttp\Psr7\stream_for() in <em class="placeholder">CRM_Shimmy_Page_FooBar-&gt;run()</em> (line <em class="placeholder">7</em> of <em class="placeholder">/home/totten/bknix/build/dmaster/web/sites/default/files/civicrm/ext/example-mixin/CRM/Shimmy/Page/FooBar.php</em>).</div>
                    <div class="clearfix">
            The website encountered an unexpected error. Please try again later.          </div>
          <div id="footer"></div>
      </div></div></div></div> <!-- /.left-corner, /.right-corner, /#squeeze, /#center -->
```


It fails because `guzzlehttp/psr7@2.0.0` [removed "all functions in the `GuzzleHttp\Psr7` namespace"](https://github.com/guzzle/psr7/blob/2.7/CHANGELOG.md#200beta-1---2021-03-21). (*Deprecated in v1.9; removed in v2.0.*)

As of `guzzlehttp/psr7` 1.9+, `stream_for()` is replaced by `Utils::streamFor()`.

Pushed up a commit to use the 1.9+ convention.

totten commented:
(*Rebased for merge-conflict*)

eileenmcnaughton commented:
oh wow thanks for that @totten  - I feel like that was a gotcha in waiting - ie sites running more recent guzzle in their CMS might have hit that (whereas it should not be a problem the other way round from my reading as the new function was not new)

eileenmcnaughton commented:
@totten I think I need to push the full guzzle update back in here - ie this got pulled back to just the psr patch experimentally -

eileenmcnaughton committed:
Composer update guzzle

docker@civicrm:/srv/civi-sites/dmaster/web/sites/all/modules/civicrm$ composer update -W guzzlehttp/guzzle
Gathering patches for root package.
Loading composer repositories with package information
Updating dependencies
Lock file operations: 0 installs, 3 updates, 0 removals
  - Upgrading guzzlehttp/guzzle (7.8.2 => 7.9.2)
  - Upgrading symfony/polyfill-mbstring (v1.27.0 => v1.30.0)
  - Upgrading symfony/polyfill-php72 (v1.27.0 => v1.30.0)

totten commented:
> I think I need to push the full guzzle update back in here - ie this got pulled back to just the psr patch experimentally -

@eileenmcnaughton Yup, go for it.

New test run hasn't fully finished -- but with this revision, the `mixin` part does pass.

The dependencies between guzzlehttp/guzzle and guzzlehttp/psr7 do have a kind of gnarly shape. But the good news: I haven't been able to find anything else in `contrib` using these utilities.

eileenmcnaughton commented:
@totten that was the only part that seemed to be failing

If we were sure that all the CMS ship with a `psr` library we could avoid having 1 in civi - but we would need to figure out standalone....

seamuslee001 commented:
@eileenmcnaughton I don't think we can be confident of that and also how Backdrop / Drupal et al might do PSR would likely be different to WP. Also as we found out that Joomla 3 is different to Joomla 4 and again different to Joomla 5 when it comes to versions of libraries and I'm not sure any of the joomla core supplies psr7 library IMO

seamuslee001 commented:
FWIW in Joomla 4 they used nyholm/psr7 for PSR7 matters https://github.com/joomla/joomla-cms/blob/4.4-dev/composer.lock#L2870 and then in Joomla 5 they are using Guzzlehttp/psr7:2.7 version https://github.com/joomla/joomla-cms/blob/5.3-dev/composer.lock#L664

totten commented:
> If we were sure that all the CMS ship with...

For purposes of the changes here, I don't think we need that level of sureness. The relevant PHP calls here are only used in tests.

* For Civi testing,  WP/BD/D7/D9/D10 builds all need to end up loading `guzzlehttp/psr7` with `^1.9 || ^2.0`.
* For Civi runtime, the Civi side should be less picky about what version is actually loaded (*as long as a singular version is coherently loaded... which is a problem for Civi-Joomla... but AFAIK it's a status-quo SNAFU with or without this PR*).

eileenmcnaughton commented:
@totten no I mean the http_discovery is kinda the idea that you can use 'whatever PSR-7, PSR-17 and PSR-18 implementations that your users have installed' https://github.com/php-http/discovery - so we could not ship guzzle & use that instead if we were confident that there was always one available - which would be true in the case of Joomla! per @seamuslee001 comments above & probably true for WP, drupal but I don't know if we are 85% or 100% sure of that

seamuslee001 commented:
@eileenmcnaughton I don't think WP core ships with a PSR17 library i think it depends on other plugins so I wouldn't be confident about WP

eileenmcnaughton commented:
@seamuslee001 Oh well it looks like this one is passing so I guess at least there is that - it should be potentially helpful to Joomla!5 users in theory as it would get them onto the same psr version

eileenmcnaughton commented:
thanks @seamuslee001 - will be interesting to see how edge goes now

agileware-justin commented:
Lived experienced here.

On WP sites, CiviCRM 5.75.1 recaptcha has been failing validations, a **false negative**.

When I traced this it came down to composer autoload and newer Guzzle libraries being bundled. CiviCRM's recaptcha called the new Guzzle library and the request fails. Recaptcha validation fails which is then shown to the end user as an invalid recaptcha. Form is not processed.

Agileware Ref: CIVICRM-2268

eileenmcnaughton commented:
@agileware-justin so you have an older library bundled in a WP extension & a newer one in Civi? Or vice versa?

agileware-justin commented:
For CiviCRM 5.75.1 recaptcha failures, found the problem earlier this week. Have had time now to submit the PR https://github.com/civicrm/civicrm-core/pull/31011

Turns out was not a Guzzle issue, although when debugging on a few different sites, I can see many multiple versions of Guzzle being bundled by the various WordPress plugins on the file system. There was only one CiviCRM extension which did this, Google Group integration by Greenleaf. The class loader scans the entire file system, will load the first class it finds that matches the class name.

I think on WordPress, most plugins tend to "get along" because they release frequently, more likely to be using a recent release or a version that is not too old - but still potentially breaky.

CiviCRM could potentially face issues here too by bundling and relying on Class libraries that are too old. Other plugins and extensions, can introduce newer versions and may break CiviCRM.

When debugging, you can see that auto loaders are executed in alphabetical order, so all those with A and B file names will execute before C (civicrm). Good choice of name for afform, I guess :)

eileenmcnaughton commented:
@agileware-justin ok - we don't have a big-picture solution for guzzle with Civi because of the multiple CMS we use but it's helpful to know that we are are more likely to cause problems if our guzzle version is too old than if it is too new

agileware-justin commented:
@eileenmcnaughton it's a tricky problem for sure. Did some quick research and looks like developers are solving this by using vendor prefixed class names.

I have gone ahead and raised a CiviCRM issue about this to see what others think, feel free to comment on it. See https://lab.civicrm.org/dev/core/-/issues/5423
